File: gb-checker/gb-checker.c
#include <stdio.h>

#define CGB_INDICATOR_BYTE 0x0143

// This magic byte address signifies which versions of Game Boy consoles can
// play a specific game.
int get_magic_byte(char *filename) {
    FILE *fp;
    int magic_byte;
    fp = fopen(filename, "r");
    fseek(fp, CGB_INDICATOR_BYTE, SEEK_SET);
    magic_byte = fgetc(fp);
    fclose(fp);
    return magic_byte;
}

int main(int argc, char *argv[]) {

    int magic_byte;

    for (int i = 1; i < argc; i++) {
        magic_byte = get_magic_byte(argv[i]);
        if (magic_byte == 0xC0) {
            printf("CGB only     (0x%.2X): ", magic_byte);
        } else if (magic_byte == 0x80) {
            printf("CGB enhanced (0x%.2X): ", magic_byte);
        } else {
            printf("GB only      (0x%.2X): ", magic_byte);
        }
        printf("%s\n", argv[i]);
    }
    return 0;
}

File: gb-checker/gb-checker.go
package main

import (
	"flag"
	"fmt"
	"log"
	"os"
)

type fileData struct {
	Filename      string
	Indicatorbyte byte
}

const cgbIndicatorByte = 0x0143

func getMagicByte(filename string) byte {
	var (
		buffer    []byte = make([]byte, 1)
		magicByte byte
	)
	file, err := os.Open(filename)
	if err != nil {
		log.Fatalf("Unable to open file: %s\n", filename)
	}
	defer file.Close()

	// ReadAt reads only len(buffer) amount of characters from a file. Neat!
	file.ReadAt(buffer, cgbIndicatorByte)
	magicByte = buffer[0]
	return magicByte
}

func main() {
	var (
		magicByte   byte
		gbOnly      []fileData
		cgbEnhanced []fileData
		cgbOnly     []fileData
	)

	flag.Parse()

	for _, filename := range flag.Args() {
		magicByte = getMagicByte(filename)
		switch magicByte {
		case 0xC0:
			cgbOnly = append(cgbOnly, fileData{filename, magicByte})
		case 0x80:
			cgbEnhanced = append(cgbEnhanced, fileData{filename, magicByte})
		default:
			gbOnly = append(gbOnly, fileData{filename, magicByte})
		}
	}
	for _, data := range cgbOnly {
		fmt.Printf("CGB only     (0x%.2X): %s\n", data.Indicatorbyte, data.Filename)
	}
	for _, data := range cgbEnhanced {
		fmt.Printf("CGB enhanced (0x%.2X): %s\n", data.Indicatorbyte, data.Filename)
	}
	for _, data := range gbOnly {
		fmt.Printf("GB only      (0x%.2X): %s\n", data.Indicatorbyte, data.Filename)
	}
}
